Native platform modifies exit process

qiuguohua opened the pull request:
Re: #

### Changelog

*  https://github.com/cocos/cocos-engine/issues/13834
https://github.com/cocos/cocos-engine/issues/12505
https://github.com/cocos/cocos-engine/issues/12545
https://github.com/cocos/cocos-engine/issues/13889
https://github.com/cocos/3d-tasks/issues/8696

![1675749494841](https://user-images.githubusercontent.com/5205637/217160848-1e56fd25-4be3-4356-aad8-ed4c428602f5.jpg)

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!
- [ ] Your pull request title is using English, it's precise and appropriate.
- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.
- [ ] Document new code with comments in source code based on API docs
- [ ] Make sure any runtime log information in `log` , `error` or `new Error('')` has been moved into `EngineErrorMap.md` with an ID, and use `logID(id)` or `new Error(getError(id))` instead.
- To official teams:
  - [ ] Check that your PR is following our [guides](https://github.com/cocos/3d-tasks/blob/master/workflows/readme.md)
-->

qiuguohua committed:
Add exit events

# Conflicts:
#	cocos/game/game.ts
#	native/cocos/platform/SDLHelper.h

qiuguohua committed:
mac and ios implementation exit process

# Conflicts:
#	native/cocos/platform/ios/IOSPlatform.mm
#	native/cocos/platform/mac/MacPlatform.h
#	native/cocos/platform/mac/MacPlatform.mm
#	native/cocos/renderer/pipeline/custom/NativeFactory.cpp

qiuguohua committed:
Modify the exit process of ios

qiuguohua committed:
Add a comment
Remove redundant code

qiuguohua committed:
Modify manual exit to request exit

qiuguohua committed:
Remove redundant code

qiuguohua committed:
Remove redundant code

qiuguohua committed:
linux and empty platform exit interface modification

qiuguohua committed:
Add omitted interface declaration

qiuguohua committed:
Fix CI errors

dumganhar commented on cocos/game/game.ts:
```diff
@@ -1032,6 +1039,12 @@ export class Game extends EventTarget {
+    private _onClose () {
+        this.emit(Game.EVENT_CLOSE);
+        // TODO : Release Resources.
+        systemInfo.exit();
```
It's strange to add a behavior `exit` to a `Info` class. 
I think `systemInfo` classes should only provide `system informations`, not to control `system`.

qiuguohua replied:
There is 'close' inside systemInfo to close the window.
And there are messages received from the system, including onClose, onShow, onHide.
So I think systemInfo should be renamed as system.

dumganhar replied:
Yep, the name should be considered carefully. `...Info` is not suitable here now.

dumganhar commented on native/cocos/platform/interfaces/modules/ISystemWindowManager.h:
```diff
@@ -55,7 +55,7 @@ class ISystemWindowManager : public OSInterface {
      * Process messages at the PAL layer
      */
-    virtual void processEvent(bool *quit) = 0;
+    virtual void processEvent(void) = 0;
```
Our code style:  (void) -> ()

dumganhar commented on native/cocos/platform/ios/IOSPlatform.h:
```diff
@@ -44,6 +44,10 @@ class IOSPlatform : public UniversalPlatform {
      */
     int32_t run(int argc, const char **argv) override;
+    
+    void requestQuit();
```
Quit or Exit ?
The aforementioned TS code call it `exit`.

qiuguohua replied:
The process is like this: close（ts）-》quit->onClose（ts）->exit
This interface is for the user to perform a manual exit。
Would it be more confusing to use exit?

dumganhar replied:
I meant how about renaming `requestQuit` to `requestExit`, not just name it `exit`.

dumganhar commented on native/cocos/platform/ios/IOSPlatform.h:
```diff
@@ -44,6 +44,10 @@ class IOSPlatform : public UniversalPlatform {
+    
+    void requestQuit();
+    
+    void exitLoop() override;
```
And the code here also calls it `exit`.

dumganhar commented on native/cocos/platform/ios/IOSPlatform.mm:
```diff
@@ -150,6 +160,25 @@ - (void)renderScene:(id)sender {
+        float sleepTime = 1000.0 / fps;
+        while (!_quitLoop) {
+            runTask();
+            usleep(sleepTime);
```
While in `exit` progress,  is `sleep` needed here?  Should it exit ASAP ?

qiuguohua replied:
If the exit is not rendering related, I'll remove

minggo replied:
I think can remove it if not meet issues.

qiuguohua replied:
OK

dumganhar commented on native/cocos/platform/ios/modules/SystemWindow.mm:
```diff
@@ -46,9 +47,10 @@ of this software and associated engine source code (the "Software"), a limited,
 }
 void SystemWindow::closeWindow() {
     // Force quit as there's no API to exit UIApplication
-    cc::events::Close::broadcast();
```
Why delete `Close::broadcast();`?

qiuguohua replied:
Before Close::broadcast() call I guess because the exit function is called directly to exit. 'Close::broadcast()' should be called in the release process

dumganhar commented on native/cocos/platform/linux/modules/SystemWindowManager.cpp:
```diff
@@ -3,23 +3,23 @@
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
```
Why change the license ?

qiuguohua replied:
I did not change it, probably due to some wrong operation.

dumganhar commented on native/cocos/bindings/auto/jsb_cocos_auto.cpp:
```diff
@@ -4616,7 +4616,7 @@ static bool js_cc_ISystemWindowManager_processEvent(se::State& s)
     ok &= sevalue_to_native(args[0], &arg2, s.thisObject());
     SE_PRECONDITION2(ok, false, "Error processing arguments"); 
-    (arg1)->processEvent(arg2);
+    (arg1)->processEvent();
```
arg2 is still defined. Did you re-generate the `auto binding code`?

qiuguohua replied:
Will this be automatically overwritten? Shouldn't this file theoretically be submitted?

dumganhar replied:
Yes ,it will be overrode by CocosRobot, but that need to send a PR. You could commit this file since the file generated by swig is consistent.

dumganhar commented on native/cocos/platform/mac/AppDelegate.mm:
```diff
@@ -104,6 +104,12 @@ - (NSWindow*)getWindow {
 
+- (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication *)sender
+{
+    //return NSTerminateCancel;
```
Remove unused comments.

qiuguohua replied:
OK

dumganhar commented on native/cocos/platform/win32/modules/SystemWindowManager.cpp:
```diff
@@ -35,21 +35,18 @@ int SystemWindowManager::init() {
 }
 
-void SystemWindowManager::processEvent(bool *quit) {
+void SystemWindowManager::processEvent(void) {
```
(void) -> ()

qiuguohua replied:
OK

qiuguohua committed:
Change quit to exit
Change exitLoop to loop

qiuguohua committed:
Modify comments

qiuguohua committed:
Fix CI error

minggo commented on @types/pal/system-info.d.ts:
```diff
@@ -48,8 +48,14 @@ declare module 'pal/system-info' {
+        /*
+         * Free the native's resources
+         */
+        public exit(): void;
```
What's the difference between close and exit?

qiuguohua replied:
'close' means close the window
'exit' means really exit the program
![image](https://user-images.githubusercontent.com/5205637/217771563-c933a56c-5571-410e-8de5-a38c22966e52.png)

minggo replied:
I don't think we should provide two methods to developers. Just one exit method is enough.

qiuguohua replied:
This interface is internal and is private in game.ts

qiuguohua replied:
We need an interface to notify native when ts release is complete

minggo replied:
Ok.

bofeng-song commented on cocos/game/game.ts:
```diff
@@ -307,6 +307,12 @@ export class Game extends EventTarget {
     public static readonly EVENT_RESUME = 'game_on_resume';
 
+    /**
+     * @en Events triggered when the game pauses to resume running. <br>
```
Whether there is any problem in the remarks

qiuguohua replied:
Yes, I remember that it was modified before, it should be an error when merging

bofeng-song commented on native/cocos/platform/SDLHelper.cpp:
```diff
@@ -223,9 +223,6 @@ void SDLHelper::dispatchSDLEvent(uint32_t windowId, const SDL_Event &sdlEvent, b
 
     switch (sdlEvent.type) {
         case SDL_QUIT: {
```
Whether SDL_QUIT is the last message to be processed in the message loop

qiuguohua replied:
SDL_QUIT is the last event customized by SDL, which can be handled or not according to the requirements

qiuguohua committed:
Modify comments
Remove sleep from running Task when IOS exits

qiuguohua commented:
https://github.com/cocos/cocos-engine-external/pull/335

qiuguohua committed:
1、Adding ohos adaptations
2、Modify third-party libraries
3、Increase the frame rate to make the program exit as soon as possible
